plugins/net_plugin/net_plugin.cpp

```diff
@@ -79,13 +80,12 @@ namespace eosio {
       std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::milliseconds(config::block_interval_ms)).count();
 
    struct node_transaction_state {
-      transaction_id_type id;
-      time_point_sec  expires;        /// time after which this may be purged.
-      uint32_t        connection_id = 0;
+      transaction_id_type  id;
+      uint32_t             connection_id = 0;
+      time_point_sec       expires;              // time after which this may be purged.
+      mutable bool         have_trx = false;     // trx received, not just trx notice, mutable because not indexed
```

Comment:
iiuc, the `have_trx` member is specific to our node, not to the pair `<id, connection_id>`, and as such I feel it should not be stored here, but maybe is a separate `unordered_map<transaction_id_type, bool>`.

I think this would simplify the code in `add_peer_txn`.

Reply:
Seems like too much overhead to store the trx id again. Also at some point we might want to know which connection the the trx was received on. At that point we could make the bool an uint8 enum with recieved_on_connection and received_on_other_connection and not_received.

Reply:
> at some point we might want to know which connection the the trx was received on

We know that we received the trx or the notice just from the fact that the pair `<id, connection_id>` is present.

> Seems like too much overhead to store the trx id again.

Hum, I don't feel this way. We already store the trx id for every connection that we received the trx or notice on. What's one more?

Reply:
It is interesting that the `connection_id` is a continuously increasing index. If we re-used the indices of old connections (we could keep a `std::vector<uint32_t> old_connection_ids;`), then we could use a dynamic_bitset instead of the `std::unordered_set`

Reply:
The `unordered_set` would work too, if we go that route I think the `boost::unordered_flat_set` would be a better choice (faster and more memory efficient, esp. for small values).